src/asyncresults.jl

```diff
@@ -0,0 +1,242 @@
+"An asynchronous PostgreSQL query"
+mutable struct AsyncResult
+    "The LibPQ.jl Connection used for the query"
+    jl_conn::Connection
+
+    "Whether this AsyncResult has locked the Connection"
+    conn_locked::Atomic{Bool}
+
+    "Keyword arguments to pass to Result on creation"
+    result_kwargs::Ref
+
+    "Task which errors or returns a LibPQ.jl Result which is created once available"
+    result_task::Task
+
+    function AsyncResult(jl_conn::Connection, conn_locked::Atomic{Bool}, result_kwargs::Ref)
+        async_result = new(jl_conn, conn_locked, result_kwargs)
+        jl_conn.async_result = async_result
+        return async_result
+    end
+end
+
+function AsyncResult(jl_conn::Connection, conn_locked::Atomic{Bool}; kwargs...)
+    AsyncResult(jl_conn, conn_locked, Ref(kwargs))
+end
+
+function AsyncResult(jl_conn::Connection; kwargs...)
+    AsyncResult(jl_conn, Atomic{Bool}(true); kwargs...)
+end
+
+function Base.show(io::IO, async_result::AsyncResult)
+    status = if isready(async_result)
+        if iserror(async_result)
+            "errored"
+        else
+            "finished"
+        end
+    else
+        "in progress"
+    end
+    print(io, typeof(async_result), " (", status, ")")
+end
```

Comment:
Should we have `status(::AsyuncResult)` as its own function?
Maybe returning an Enum?

or is that over-engineered?

Reply:
I think in programming you usually only care about `isready` and then `iserror`; I can't see `status` being additionally useful for anything other than this function.